### PR TITLE
Add utilities for modular arithmetic

### PR DESCRIPTION
**Pull request by chiphogg**
These utilities all operate on unsigned 64-bit integers.  The main
design goal is to produce correct answers for all inputs while avoiding
overflow.

Most operations are standard: addition, subtraction, multiplication,
powers.  The one non-standard one is "half_mod_odd".  This operation is
`(x/2) % n`, but only if `n` is odd.  If `x` is also odd, we add `n`
before dividing by 2, which gives us an integer result.  We'll need this
operation for the strong Lucas probable prime checking later on.

These are Au-internal functions, so we use unchecked preconditions: as
long as the caller makes sure the inputs are already reduced-mod-n,
we'll keep them that way.

Helps #217.

### CODE REVIEW

File: au/code/au/utility/mod.hh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include <limits>
+
+namespace au {
+namespace detail {
+
+// (a * b) % n
+//
+// Precondition: (a < n).
+// Precondition: (b < n).
+constexpr uint64_t add_mod(uint64_t a, uint64_t b, uint64_t n) {
```
Comment (geoffviola) [inline au/code/au/utility/mod.hh:27]: `((3 * 2) % 4)` is 2, but `add_mod(3, 2, 4)` is 1. Is that expected?

**Review by geoffviola — CHANGES_REQUESTED** [review]

File: au/code/au/utility/mod.hh (hunk 1 above)
Comment (chiphogg) [reply au/code/au/utility/mod.hh:27]: Yes, because `((3 + 2) % 4)` is `1`.  :wink:

(I expect you'll find that `mul_mod(3, 2, 4)` gives `2`.)

Comment (chiphogg) [reply au/code/au/utility/mod.hh:27]: Ah, I see --- documentation fail.  Let me go fix the comment...

Comment (chiphogg) [reply au/code/au/utility/mod.hh:27]: Fixed.  Also, thought of a new test case, and found one last overflow bug!  This one was only possible to hit when testing the max int $(2^{64} - 1)$.

File: au/code/au/utility/mod.hh [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    } else {
+        return n - (b - a);
+    }
+}
+
+// (a * b) % n
+//
+// Precondition: (a < n).
+// Precondition: (b < n).
+constexpr uint64_t mul_mod(uint64_t a, uint64_t b, uint64_t n) {
```
Comment (geoffviola) [inline au/code/au/utility/mod.hh:51]: This implementation seems right. It gets the right answer for the `JUST_UNDER_HALF, 10u, MAX` case as opposed to https://www.geeksforgeeks.org/how-to-avoid-overflow-in-modular-multiplication/

**Review by geoffviola — APPROVED** [review]
Implementation looks good. Tests look good. Thanks for the comments.
